libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp

```diff
@@ -220,470 +233,658 @@ void AP_Frsky_Telem::sport_tick(void)
 
-/*
- * prepare sats information stored in member variables
- */
-void AP_Frsky_Telem::calc_gps_sats()
-{
-    // GPS status is sent as num_sats*10 + status, to fit into a uint8_t
-    const AP_GPS &gps = _ahrs.get_gps();
-    gps_sats = gps.num_sats() * 10 + gps.status();
+    if (paramID >= 4) {
+        paramID = 0;
+    }
+    paramID++;
+    switch(paramID) {
+    case 1:
+        param = _params.mav_type;                                    // frame type (see MAV_TYPE in Mavlink definition file common.h)
+        break;
+    case 2:
+        param = (uint32_t)roundf(*_params.fs_batt_voltage * 100.0f); // battery failsafe voltage in centivolts
+        break;
+    case 3:
+        param = (uint32_t)roundf(*_params.fs_batt_mah);              // battery failsafe capacity in mAh
+        break;
+    case 4:
+        param = (uint32_t)roundf(_battery.pack_capacity_mah());      // battery pack capacity in mAh as configured by user
+        break;
+    }
+    //Reserve first 8 bits for param ID, use other 24 bits to store parameter value
+    param = (paramID << 24) | (param & 0xFFFFFF);
+    
+    return param;
 }
 
 /*
- * send number of gps satellite and gps status eg: 73 means 7 satellite and 3d lock
+ * prepare gps latitude/longitude data
+ * for FrSky SPort Passthrough (OpenTX) protocol (X-receivers)
  */
-void AP_Frsky_Telem::send_gps_sats()
+uint32_t AP_Frsky_Telem::calc_gps_latlng(bool *send_latitude)
 {
-    frsky_send_data(FRSKY_ID_TEMP2, gps_sats);
-}
+    uint32_t latlng = 0;
+    const Location &loc = _ahrs.get_gps().location(0); // use the first gps instance (same as in send_mavlink_gps_raw)
 
-/*
- * send control_mode as Temperature 1 (TEMP1)
- */
-void AP_Frsky_Telem::send_mode(void)
-{
-    frsky_send_data(FRSKY_ID_TEMP1, _mode);
+    // alternate between latitude and longitude
+    if (*send_latitude == true) {
+        if (loc.lat < 0) {
+            latlng |= ((abs(loc.lat)/100)*6) | 0x40000000;
+        } else {
+            latlng |= ((abs(loc.lat)/100)*6);
+        }
+        *send_latitude = false;
+    } else {
+        if (loc.lng < 0) {
+            latlng |= ((abs(loc.lng)/100)*6) | 0xC0000000;
+        } else {
+            latlng |= ((abs(loc.lng)/100)*6) | 0x80000000;
+        }
+        *send_latitude = true;
+    }
+    return latlng;
 }
 
 /*
- * send barometer altitude integer part . Initialize baro altitude
+ * prepare gps status data
+ * for FrSky SPort Passthrough (OpenTX) protocol (X-receivers)
  */
-void AP_Frsky_Telem::send_baro_alt_m(void)
+uint32_t AP_Frsky_Telem::calc_gps_status(void)
 {
-    frsky_send_data(FRSKY_ID_BARO_ALT_BP, _baro_alt_meters);
-}
+    uint32_t gps_status = 0;
 
-/*
- * send barometer altitude decimal part
- */
-void AP_Frsky_Telem::send_baro_alt_cm(void)
-{
-    frsky_send_data(FRSKY_ID_BARO_ALT_AP, _baro_alt_cm);
+    // number of GPS satellites visible
+    if (_ahrs.get_gps().num_sats() > 15) { // limit number of sats to 15
+        gps_status |= 15;
```

Comment:
There are a lot of magical numbers, I guess that's what you get when people are trying to protect a proprietary protocol. But in this case I don't see it as magical, 15 is the maximum number of satellites they support to be sent. Sure there might be better ways to write it, but it isn't wrong.